### ISSUE TITLE
hm_messages.messageuid uniqueness?

### ISSUE THREAD

**Issue by SorenRR** [hmailserver/hmailserver#475]
Been tracing the HM5136 error for a while on my 5.6.9 RvdH-fork + minor personal mods.

"ERROR"	2468	"2023-09-12 12:10:24.688"	"Severity: 3 (Medium), Code: HM5136, Source: TCPConnection::AsyncReadCompleted, Description: An error occured while parsing data. Data length: 48, Data: A311 UID FETCH 577985 (BODY.PEEK[]<43911.65536>). Remote IP: 192.168.0.61"
"ERROR"	2468	"2023-09-12 12:10:25.032"	"Severity: 2 (High), Code: HM4208, Source: ExceptionHandler::Run, Description: An error occured while executing 'IOCPQueueWorkerTask'"
"ERROR"	2468	"2023-09-12 12:10:27.250"	"Severity: 1 (Critical), Code: HM5519, Source: StackLogger::Log, Description: An error has been detected. A mini dump has been written to C:\hMailServer\Logs\minidump_2023-09-12 121025_{D43295B9-19AD-404B-98EC-908B074AFDAC}.dmp"

Now the thing is I ran:

SELECT hm_messages.messageuid, hm_accounts.accountaddress, hm_messages.messagefilename, hm_messages.messageid FROM hm_messages
INNER JOIN hm_accounts ON hm_accounts.accountid = hm_messages.messageaccountid
WHERE hm_messages.messageuid = 577985 ;

It produced a set of duplicates, one listing my mailbox and the other listing a different mailbox. The error came from my mailbox while opening the message (IMAP) but the first result from the select statement (unsorted) listed the other mailbox and I presume that is why the BODY.PEEK[]<43911.65536> failed as the message size differed.

My database dates back to 2006 and the older the data, the bigger the number of duplicates.

Does all this make sense or have I got it completely wrong?

**Comment by SorenRR** [comment]
Both email files exist.

![emails](https://github.com/hmailserver/hmailserver/assets/8338115/6af148c4-ce91-4b8f-8f57-69f8421ad008)

**Comment by RvdHout** [comment]
Don't think unless you find and fix the issue yourself posting here is of much help.
IMAP, and especially FETCH / PEEK seems to have quitte a few flaws and bugs

https://github.com/hmailserver/hmailserver/issues/334
https://github.com/hmailserver/hmailserver/issues/459

Both these are obvious and confirmed bugs, and nothing is (unfortunately) really done with it

**Comment by SorenRR** [comment]
Yeah, I noticed that a lot of this code stuff is changed in 5.7

**Comment by RvdHout** [comment]
Is it? code looks very similar to me, and at least i have the same bug #459 in both 5.6.x. and 5.7.x
To pinpoint the issue you could check if fetched data matches sizes given (fetched, peeked)

**Comment by SorenRR** [comment]
Compare IMAPCommandExpunge.cpp (master ~ line 74, 5.6.9 ~ line 55)

If there is a problem expunging that would explain why 95% of the HM5136 problems I have is with Apple gear. Apple typically "mark for delete" but never deletes where as all the other clients I have will delete email on the spot.

Something is weird here and I cannot see there is any simple way to fix up the database to eliminate existing duplicate messageuid's.

The easiest thing would be to create a unique index on messageuid but for that to work the database must be cleaned first. Not sure backing up and restoring will fix it. With a unique index the bug will reveal itself ;-)

**Comment by SorenRR** [comment]
OK the more I read about this UID value the more it sucks!
UID: https://datatracker.ietf.org/doc/html/rfc3501#section-2.3.1.1
Then there is an IMAP thing called UIDVALIDITY wich together with UID should uniquely identify the message in the mailbox (IMAP slang for FOLDER) I guess that is why there is a "foldercurrentuid" column in  table hm_IMAPFolders.
The IMAP UID of a message can be system-wide unique or mailbox-wide unique. System-wide I expect a very large number (as in this case) or mailbox-wide a very small number < 1,000. Also to add to the confusion is POP3 UIDL which is a system-wide identifier. In rare cases mailservers use the SAME UID/UIDL for both IMAP and POP3.

Anyways, Primary focus for me now is 1) to fix the error message for HM5136 to include ALL (!) relevant information to support debugging (account & "mailbox") and 2) IS there really a problem with "real" duplicate UID's or is the problem somewhere else as RvdH suggests with the BODY.PEEK[]<43911.65536>

**Comment by SorenRR** [comment]
Got a new one. Loaded the minidump into Visual Studio...

`2524		"2023-09-26 17:26:01.800"	""
2524		"2023-09-26 17:26:01.800"	"iSeverity: 3"
2524		"2023-09-26 17:26:01.800"	"iCode: 5136"
2524		"2023-09-26 17:26:01.800"	"sSource: TCPConnection::AsyncReadCompleted"
2524		"2023-09-26 17:26:01.800"	"sDescription: An error occured while parsing data. Data length: 48, Data: A309 UID FETCH 578067 (BODY.PEEK[]<43934.65536>). Remote IP: 192.168.0.61"
2524		"2023-09-26 17:26:01.832"	"MessageUID      = 578067"
2524		"2023-09-26 17:26:01.832"	"Start (octet)   = 43934"
2524		"2023-09-26 17:26:01.832"	"Length (octet)  = 65536"
2524		"2023-09-26 17:26:01.832"	"IPAddress       = 192.168.0.61"
2524		"2023-09-26 17:26:01.894"	"accountaddress   = jane@home.arpa"
2524		"2023-09-26 17:26:01.894"	"messagefilename  = {44D232A4-FF2F-4EBF-BF23-BC70398B91B6}.eml"
2524		"2023-09-26 17:26:01.894"	"strFile          = C:\hMailServer\Data\home.arpa\jane\44\{44D232A4-FF2F-4EBF-BF23-BC70398B91B6}.eml"
2524		"2023-09-26 17:26:01.910"	"C:\hMailServer\Data\home.arpa\jane\44\{44D232A4-FF2F-4EBF-BF23-BC70398B91B6}.eml = 104938 bytes"
2524		"2023-09-26 17:26:01.910"	"accountaddress   = soren@home.arpa"
2524		"2023-09-26 17:26:01.910"	"messagefilename  = {1583562C-0BDC-4CDC-8277-B6C68D8A7915}.eml"
2524		"2023-09-26 17:26:01.910"	"strFile          = C:\hMailServer\Data\home.arpa\soren\15\{1583562C-0BDC-4CDC-8277-B6C68D8A7915}.eml"
2524		"2023-09-26 17:26:01.910"	"C:\hMailServer\Data\home.arpa\soren\15\{1583562C-0BDC-4CDC-8277-B6C68D8A7915}.eml = 43934 bytes"
2524		"2023-09-26 17:26:01.941"	""
2524		"2023-09-26 17:26:01.941"	"iSeverity: 2"
2524		"2023-09-26 17:26:01.941"	"iCode: 4208"
2524		"2023-09-26 17:26:01.941"	"sSource: ExceptionHandler::Run"
2524		"2023-09-26 17:26:01.941"	"sDescription: An error occured while executing 'IOCPQueueWorkerTask'"
2524		"2023-09-26 17:26:03.410"	""
2524		"2023-09-26 17:26:03.410"	"iSeverity: 1"
2524		"2023-09-26 17:26:03.410"	"iCode: 5519"
2524		"2023-09-26 17:26:03.410"	"sSource: StackLogger::Log"
2524		"2023-09-26 17:26:03.410"	"sDescription: An error has been detected. A mini dump has been written to C:\hMailServer\Logs\minidump_2023-09-26 172601_{3CB98C40-2E76-40F7-8116-7B13B891E58F}.dmp"`

Debugger says:

Unhandled exception at 0x00569004 (hMailServer.exe) in minidump_2023-09-26 172601_{3CB98C40-2E76-40F7-8116-7B13B891E58F}.dmp: 0xC0000005: Access violation reading location 0x00000004.

Fileutilities.cpp -- line 269 -- memcpy(OutBuf, bytes->GetBuffer(), iCount);

**Comment by hunterius-prime** [comment]
I am not sure if there is comment here helps (hopefully does).
We are tracking this error for more than week, testing an parsing.  Now we are able to reproduce this error.
Test case:  If you have message with attachment ( multipart) that has size 300 000B and you send 
UID FETCH 13060 (UID BODY.PEEK[2]<0.393216>)
server answer is correct
* 149 FETCH (UID 13060 BODY[2]<0> {300000}

but IOS device mail often after that send
UID FETCH 13060 (UID BODY.PEEK[2]<393216.393216>)
what is wrong command but it doesn't matter

this cases crash on parsing BadAlocation , because it tries to read outside of buffer.
"HM4208, Source: ExceptionHandler::Run, Description: An error occured while executing 'IOCPQueueWorkerTask', Message: bad allocation"
this error is consistent in versions 5.6.6, 5.6.7, 5.6.8, 5.6.9, 5.6.9-mod

I was able to find probable spot in code but sadly unable to compile hmailserver :-(

**Comment by RvdHout** [comment]
What is/are your issue(s) with compiling hmailserver?
Every branch has a detailed guide how to build that specific branch listed under "Building hMailServer"

**Comment by hunterius-prime** [comment]
> What is/are your issue(s) with compiling hmailserver? Every branch has a detailed guide how to build that specific branch listed under "Building hMailServer"
> 
> 5.6.9 https://github.com/hmailserver/hmailserver/blob/5.6.9/README.md
> 
> 5.7.x https://github.com/hmailserver/hmailserver/blob/master/README.md
> 
> If you encounter specific issues building, please name them, maybe i (or others) can help you solve those

right now I was focused on reproducing error, my colleague was trying to prepare environment ( machine) for building.
Main issues are with old development tools ( we are not C++ devs but using daily VS 2022 but find and install VS2013 is hard enough) , right now he is in third wave of preparing environment , and failing even to build boost 
(further we got was compiling hmailserver and got error of finding "libboost_date_time-vc120-mt-gd-x32-1_70.lib")

our idea is find and fix error under your version of 5.6.9 . I am going to prepare clean machine during weekend and will try to get back with problem with compiling to another issue as this is dedicated to specific error/problem

**Comment by RvdHout** [comment]
Here is complete download VS2013 community update 5 edition iso if you need it
https://download.microsoft.com/download/A/A/D/AAD1AA11-FF9A-4B3C-8601-054E89260B78/vs2013.5_ce_enu.iso

OpenSSL 1.1.1x
https://github.com/openssl/openssl/releases/tag/OpenSSL_1_1_1w

Boost 1.70.0
https://archives.boost.io/release/1.70.0/source/

if you follow the build instructions to the letter, you should be able to compile it just fine

**Comment by RvdHout** [comment]
Out of curiosity, what is the probable spot you are referring to? (I won't be able to test as i don't own a IOS device)

**Comment by hunterius-prime** [comment]
> Out of curiosity, what is the probable spot you are referring to? (I won't be able to test as i don't own a IOS device)

now we test it only with console app and IMAP commands , not with IOS device ( what was initial source of serious problems as we have rather big install of hmailserver with mostly IMAP clients)
if you send 
UID FETCH 13060 (UID BODY.PEEK[2]<393216.393216>) 
on part that size is lower than 393216B ( octet_start)
you will crash parsing

my blind shot is IMAPFetch.cpp  ln. 269

 // Jump forward to the start of the buffer.
 iBufferSize -= iOctetStart;
 
 if your body part size is (thus buffer size) 300 000B and octet_start is 393216
 you will jump start to -93216B
 
 but maybe there is more to it , will know more after running compilation

**Comment by hunterius-prime** [comment]
> Here is complete download for Visual Studio 2013 Community Edition (update 5) iso if you need it https://download.microsoft.com/download/A/A/D/AAD1AA11-FF9A-4B3C-8601-054E89260B78/vs2013.5_ce_enu.iso
> 
> Windows Driver Kit Version 7.1.0 ISO https://download.microsoft.com/download/4/A/2/4A25C7D5-EFBE-4182-B6A9-AE6850409A78/GRMWDK_EN_7600_1.ISO
> 
> OpenSSL 1.1.1x https://github.com/openssl/openssl/releases/tag/OpenSSL_1_1_1w
> 
> Boost 1.70.0 https://archives.boost.io/release/1.70.0/source/
> 
> if you follow the build instructions to the letter, you should be able to compile it just fine

Sadly, no luck even with build: on clean Win10 machine, followed instructions, and (after fixing some linked version mismatch) , when building in debug  I got error

error C3859: virtual memory range for PCH exceeded; please recompile with a command line option of '-Zm132' or greater

ok, then I did that, started build and while compiling SURBL.cpp

fatal error C1060: compiler is out of heap space

so far no luck with googling solution

**Comment by RvdHout** [comment]
What are you using as source?

**Comment by hunterius-prime** [comment]
your repository 5.6.9-mod

**Comment by RvdHout** [comment]
And if you build in /release mode?

**Comment by hunterius-prime** [comment]
release let me go few files further with same scenario
so basically
debug Win32  
release Win32
debug x64
release x64

all in the end ended in same error
and /Zm132 I have to use  ( if not I will end up with first error C3859: virtual memory range for PCH exceeded;)

btw machine is I7 notebook with 16GB RAM

**Comment by RvdHout** [comment]
can you try this? In hMailServer.vcxproj

line 113:
`<AdditionalOptions>/ignore:4099 /ignore:4098 /ignore:4254 /ignore:4078 %(AdditionalOptions)</AdditionalOptions>`

line 169
`<AdditionalOptions>/SAFESEH /ignore:4099 /ignore:4098 /ignore:4254 /ignore:4078 %(AdditionalOptions)</AdditionalOptions>`

**Comment by hunterius-prime** [comment]
the same behavior

**Comment by RvdHout** [comment]
Post/attach your hMailServer.vcxproj

**Comment by hunterius-prime** [comment]
please give me some time as I am going from start again ( code part)

**Comment by hunterius-prime** [comment]
[hMailServer.zip](https://github.com/user-attachments/files/17350589/hMailServer.zip)

**Comment by hunterius-prime** [comment]
`1>  SpamTestSPF.cpp
1>  SpamTestSURBL.cpp
1>  SURBL.cpp
1>  WhiteListCache.cpp
1>C:\dev\hMailLibs\boost_1_70_0\boost/type_traits/common_type.hpp(117): fatal error C1001: An internal error has occurred in the compiler.
1>  (compiler file 'f:\dd\vctools\compiler\cxxfe\sl\p1\c\esumem.c', line 8636)
1>   To work around this problem, try simplifying or changing the program near the locations listed above.
1>  Please choose the Technical Support command on the Visual C++ 
1>   Help menu, or open the Technical Support help file for more information (..\Common\AntiSpam\SURBL.cpp)
1>          C:\dev\hMailLibs\boost_1_70_0\boost/type_traits/common_type.hpp(139) : see reference to class template instantiation 'boost::type_traits_detail::common_type_impl<T1,T2>' being compiled
1>          with
1>          [
1>              T1=int_least64_t
1>  ,            T2=int_least64_t
1>          ]
1>          C:\dev\hMailLibs\boost_1_70_0\boost/type_traits/common_type.hpp(145) : see reference to class template instantiation 'boost::type_traits_detail::common_type_decay_helper<T1,T2,__int64,__int64>' being compiled
1>          with
1>          [
1>              T1=int_least64_t
1>  ,            T2=int_least64_t
1>          ]
1>          C:\dev\hMailLibs\boost_1_70_0\boost/chrono/duration.hpp(405) : see reference to class template instantiation 'boost::common_type<Rep1,Rep2>' being compiled
1>          with
1>          [
1>              Rep1=int_least64_t
1>  ,            Rep2=int_least64_t
1>          ]
1>          ..\Common\AntiSpam\SURBL.cpp(139) : see reference to class template instantiation 'boost::common_type<boost::chrono::system_clock::duration,boost::chrono::system_clock::duration>' being compiled
1>  AntiVirusConfiguration.cpp
1>  ClamAVVirusScanner.cpp
1>c1xx : fatal error C1060: compiler is out of heap space`

**Comment by hunterius-prime** [comment]
part of output:
1>  SpamTestSPF.cpp
1>  SpamTestSURBL.cpp
1>  SURBL.cpp
1>  WhiteListCache.cpp
1>C:\dev\hMailLibs\boost_1_70_0\boost/type_traits/common_type.hpp(117): fatal error C1001: An internal error has occurred in the compiler.
1>  (compiler file 'f:\dd\vctools\compiler\cxxfe\sl\p1\c\esumem.c', line 8636)
1>   To work around this problem, try simplifying or changing the program near the locations listed above.
1>  Please choose the Technical Support command on the Visual C++ 
1>   Help menu, or open the Technical Support help file for more information (..\Common\AntiSpam\SURBL.cpp)
1>          C:\dev\hMailLibs\boost_1_70_0\boost/type_traits/common_type.hpp(139) : see reference to class template instantiation 'boost::type_traits_detail::common_type_impl<T1,T2>' being compiled
1>          with
1>          [
1>              T1=int_least64_t
1>  ,            T2=int_least64_t
1>          ]
1>          C:\dev\hMailLibs\boost_1_70_0\boost/type_traits/common_type.hpp(145) : see reference to class template instantiation 'boost::type_traits_detail::common_type_decay_helper<T1,T2,__int64,__int64>' being compiled
1>          with
1>          [
1>              T1=int_least64_t
1>  ,            T2=int_least64_t
1>          ]
1>          C:\dev\hMailLibs\boost_1_70_0\boost/chrono/duration.hpp(405) : see reference to class template instantiation 'boost::common_type<Rep1,Rep2>' being compiled
1>          with
1>          [
1>              Rep1=int_least64_t
1>  ,            Rep2=int_least64_t
1>          ]
1>          ..\Common\AntiSpam\SURBL.cpp(139) : see reference to class template instantiation 'boost::common_type<boost::chrono::system_clock::duration,boost::chrono::system_clock::duration>' being compiled
1>  AntiVirusConfiguration.cpp
1>  ClamAVVirusScanner.cpp
1>c1xx : fatal error C1060: compiler is out of heap space

**Comment by RvdHout** [comment]
[hMailServer.zip](https://github.com/user-attachments/files/17350601/hMailServer.zip)

**Comment by hunterius-prime** [comment]
----- Build started: Project: hMailServer, Configuration: Debug Win32 ------
1>  The hMailServer service is not started.
1>  
1>  More help is available by typing NET HELPMSG 3521.
1>  
1>  stdafx.cpp
1>  AntiSpamConfiguration.cpp
1>  Canonicalization.cpp
1>  DKIM.cpp
1>  DKIMParameters.cpp
1>  DKIMSigner.cpp
1>  SpamTestDKIM.cpp
1>  SpamAssassinClient.cpp
1>  SpamAssassinTestConnect.cpp
1>  SpamProtection.cpp
1>  SpamTest.cpp
1>  SpamTestData.cpp
1>  SpamTestDNSBlackLists.cpp
1>  SpamTestHeloHost.cpp
1>  SpamTestMXRecords.cpp
1>  SpamTestPTR.cpp
1>  SpamTestResult.cpp
1>  SpamTestRunner.cpp
1>  SpamTestSpamAssassin.cpp
1>  SpamTestSPF.cpp
1>  SpamTestSURBL.cpp
1>  SURBL.cpp
1>  WhiteListCache.cpp
1>  AntiVirusConfiguration.cpp
1>  ClamAVVirusScanner.cpp
1>  ClamWinVirusScanner.cpp
1>C:\dev\hMailLibs\boost_1_70_0\boost/type_traits/common_type.hpp(117): fatal error C1001: An internal error has occurred in the compiler.
1>  (compiler file 'f:\dd\vctools\compiler\cxxfe\sl\p1\c\esumem.c', line 8636)
1>   To work around this problem, try simplifying or changing the program near the locations listed above.
1>  Please choose the Technical Support command on the Visual C++ 
1>   Help menu, or open the Technical Support help file for more information (..\Common\AntiSpam\SURBL.cpp)
1>          C:\dev\hMailLibs\boost_1_70_0\boost/type_traits/common_type.hpp(139) : see reference to class template instantiation 'boost::type_traits_detail::common_type_impl<T1,T2>' being compiled
1>          with
1>          [
1>              T1=int_least64_t
1>  ,            T2=int_least64_t
1>          ]
1>          C:\dev\hMailLibs\boost_1_70_0\boost/type_traits/common_type.hpp(145) : see reference to class template instantiation 'boost::type_traits_detail::common_type_decay_helper<T1,T2,__int64,__int64>' being compiled
1>          with
1>          [
1>              T1=int_least64_t
1>  ,            T2=int_least64_t
1>          ]
1>          C:\dev\hMailLibs\boost_1_70_0\boost/chrono/duration.hpp(405) : see reference to class template instantiation 'boost::common_type<Rep1,Rep2>' being compiled
1>          with
1>          [
1>              Rep1=int_least64_t
1>  ,            Rep2=int_least64_t
1>          ]
1>          ..\Common\AntiSpam\SURBL.cpp(139) : see reference to class template instantiation 'boost::common_type<boost::chrono::system_clock::duration,boost::chrono::system_clock::duration>' being compiled
1>  CustomVirusScanner.cpp
1>c1xx : fatal error C1060: compiler is out of heap space
1>  Note: non-fatal system error releasing memory (87)
1>  VirusScanner.cpp
1>c1xx : fatal error C1060: compiler is out of heap space
1>  Note: non-fatal system error releasing memory (87)
1>  VirusScannerTester.cpp
1>c1xx : fatal error C1060: compiler is out of heap space
1>  Note: non-fatal system error releasing memory (87)
1>  VirusScanningResult.cpp
1>c1xx : fatal error C1060: compiler is out of heap space
1>  Note: non-fatal system error releasing memory (87)
1>  ACLManager.cpp

**Comment by RvdHout** [comment]
You did build/compile boost libs, right? 

1. Download Boost 1.70.0 from http://www.boost.org/ and put it into %hMailServerLibs%<Boost-Version>. 
You should now have a folder named %hMailServerLibs%<Boost-Version>, for example C:\Dev\hMailLibs\boost_1_70_0
2. Start a VS2013 x86 Native Tools Command Prompt
3. Change dir to %hMailServerLibs%<Boost-Version>.
4. Run the following commands:

```
bootstrap.bat vc12
bjam.exe --toolset=msvc-12.0 --build-type=complete address-model=32 --build-dir=out32 -j 4
```

**NOTE:** Change the -j parameter from 4 to the number of cores on your computer. The parameter specifies the number of parallel compilations will be done.

**Comment by hunterius-prime** [comment]
yes, I did, with some errors in process,  and message some modules was not updated

and I also have precompiled windows binaries from  boost ( to test change)

**Comment by RvdHout** [comment]
> yes, I did, with some errors in process, and message some modules was not updated
You shouldn't, it should finish without errors

> and I also have precompiled windows binaries from boost ( to test change)
U what? So far for "please follow the instructions to the letter"

**Comment by hunterius-prime** [comment]
![image](https://github.com/user-attachments/assets/19fb5fad-a73a-4023-bb79-2bf5973c3eef)
trying again clean, will report back

**Comment by RvdHout** [comment]
Doubt it will matter, but maybe try running 'VS2013 x86 Native Tools Command Prompt' as Administrator 

![image](https://github.com/user-attachments/assets/95b1f4f9-f2b2-4561-b835-3ab8146c1aa9)

**Comment by hunterius-prime** [comment]
while waiting for boost compilation I found failing part of boost source
`#if !defined(BOOST_MSVC) || BOOST_MSVC > 1800

// internal compiler error on msvc-12.0

template<class T1, class T2> using builtin_common_type = typename boost::decay<decltype( boost::declval<bool>()? boost::declval<T1>(): boost::declval<T2>() )>::type;

template<class T1, class T2> struct common_type_impl: mp_defer<builtin_common_type, T1, T2>
{
};

#else

template<class T1, class T2> using builtin_common_type = decltype( boost::declval<bool>()? boost::declval<T1>(): boost::declval<T2>() );

template<class T1, class T2> struct common_type_impl_2: mp_defer<builtin_common_type, T1, T2>
{
};

template<class T1, class T2> using decay_common_type = typename boost::decay<typename common_type_impl_2<T1, T2>::type>::type;

template<class T1, class T2> struct common_type_impl: mp_defer<decay_common_type, T1, T2>
{
};

#endif // !defined(BOOST_MSVC) || BOOST_MSVC > 1800`

and I shouldnt even be there so probably BOOST_MSVC isn't defined or it is >1800

**Comment by RvdHout** [comment]
It really should compile without issues using default sources, **period!**

**Comment by hunterius-prime** [comment]
please, have patience 
I did everything according to instructions ( twice) , used only links you provided, without success , and start wondering only after problems . So believe me I was following instructions, and yet Boost didn't compile  successfully. Tomorrow I will start from scratch and will try not to update Win10 to latest version , but I don't believe this is the problem.

**Comment by RvdHout** [comment]
There ain't any rocket science involved, a few easy steps to follow and you should be building in no time, so obviously you did not follow the instructions to the [letter](https://github.com/RvdHout/hmailserver/blob/5.6.9-Mod/README.md), i replicated all mentioned steps, using the same tools and same instructions and i can build both debug/release builds without issues. 

Updating win10 (or not) is irrelevant to your problem.... to me it looks like you used (partially) the 5.7 build instructions (which is 64-bit) and not the build instructions for [5.6.9 ](https://github.com/RvdHout/hmailserver/blob/5.6.9-Mod/README.md)

**Comment by SorenRR** [comment]
> please, have patience I did everything according to instructions ( twice) , used only links you provided, without success , and start wondering only after problems . So believe me I was following instructions, and yet Boost didn't compile successfully. Tomorrow I will start from scratch and will try not to update Win10 to latest version , but I don't believe this is the problem.
> 
> ...failed updating 12 targets... ...skipped 96 targets... ...updated 3104 targets...

My system (HMS 5.6.9) is build according to instructions - except compiling OpenSSL as I have upgraded to version 3.0.15 - HOWEVER ... EVERYTHING else is as described, installed on Windows 10 Home running latest updates!

Also, when building BOOST lib and OpenSSL lib it is IMPERATIVE to do it from the "VS2013 x86 Native Tools Command Prompt" running as Administrator !

Re BOOST issues... I run this command on the BOOST directory as I have had some issues with line endings...

for /r %%f in (*.bat) do unix2dos %%f
for /r %%f in (*.cmd) do unix2dos %%f

**Comment by hunterius-prime** [comment]
First of all(not resolved), recapitulation for anyone who may will be stuck in the same situation: I have Win 10 Home machine, clean install updated to H22 version, with installed MSSQL Express 2022 and SSMS 20.2, Install of VS2013 will go with error "some modules not loaded correctly" but repair went well.  When building Boost it failed on compiler internal error , because all compiler directives uses  MSVC>1800 (v18.00.xxx)but in this machine MSVC simply identify as 1900 (v18.0.30723) and thus will always jump into forbidden parts of source. I don't have time  to investigate it further, neither install from scratch.

I was able to build and run VS2019 and hMailserver 5.7 ( master branch ) so we will continue with this version, because time is of the essence.

I was able to reproduce issue, and it is either simple either easy to fix, just I want to discuss what would be better solution.
Source of problem: if you request part of body through IMAP with "start" and "length" definition that is more than body part size you will end up with wrong start/length values when retrieving data.

question is:
what should be expected final behavior when sending wrong request?

**Comment by hunterius-prime** [comment]
> > please, have patience I did everything according to instructions ( twice) , used only links you provided, without success , and start wondering only after problems . So believe me I was following instructions, and yet Boost didn't compile successfully. Tomorrow I will start from scratch and will try not to update Win10 to latest version , but I don't believe this is the problem.
> > ...failed updating 12 targets... ...skipped 96 targets... ...updated 3104 targets...
> 
> My system (HMS 5.6.9) is build according to instructions - except compiling OpenSSL as I have upgraded to version 3.0.15 - HOWEVER ... EVERYTHING else is as described, installed on Windows 10 Home running latest updates!
> 
> Also, when building BOOST lib and OpenSSL lib it is IMPERATIVE to do it from the "VS2013 x86 Native Tools Command Prompt" running as Administrator !
> 
> Re BOOST issues... I run this command on the BOOST directory as I have had some issues with line endings...
> 
> for /r %%f in (_.bat) do unix2dos %%f for /r %%f in (_.cmd) do unix2dos %%f

I will return to it later, but right now I need to fix problem first and send server to testing

**Comment by hunterius-prime** [comment]
Why I am asking:  when I overshoot requested part length in first request according to RFC it will return actual part and also size what was returned. That is correct.
But IOS tend to request body part in chunks of 393216B and sometimes continue even while part ends.
so fallback could be:
- again reset vales to <0.partsize>  ( but this could lead to repeated circular request - we will test it)
- reset value to <partsize.partsize> and return nothing ( is it ok with RFC? )
- return error message? ( I really don't want to do that)

**Comment by RvdHout** [comment]
> Why I am asking: when I overshoot requested part length in first request according to RFC it will return actual part and also size what was returned. That is correct. But IOS tend to request body part in chunks of 393216B and sometimes continue even while part ends. so fallback could be:
> 
> * again reset vales to <0.partsize>  ( but this could lead to repeated circular request - we will test it)
> * reset value to <partsize.partsize> and return nothing ( is it ok with RFC? )
> * return error message? ( I really don't want to do that)

return nothing, or silently disregard... the problem seems IOS not following the RFC, i have seen a single with other (proper) IMAP clients

**Comment by RvdHout** [comment]
There is a confirmed bug however in IMAP FETCH that sometimes returns the wrong part....  https://github.com/hmailserver/hmailserver/issues/459

Not sure if it could be related, as i only can trigger/replicate that using a forward where the forwarded message is attached as message/rfc822 part ( eg: in Outlook, forward as attachment)

**Comment by hunterius-prime** [comment]
> There is a confirmed bug however in IMAP FETCH that sometimes returns the wrong part.... #459
> 
> Not sure if it could be related, as i only can trigger/replicate that using a forward where the forwarded message is attached as message/rfc822 part ( eg: in Outlook, forward as attachment)

Tomorrow we will setup testing environment with new build and transfer some big IMAP account to it so we will test it . also will look at #459

**Comment by RvdHout** [comment]
> Is it possible to get such .eml message that causes trouble?

Are you using roundcube webmail? Forwarding a message to yourself **as attachment** (within roundcube) should trigger it

An example message is attached here:
https://github.com/roundcube/roundcubemail/issues/9011

**Comment by RvdHout** [comment]
> As of now it looks like we got rid of errors HM5136, HM4208 and subsequent "OutOfMemoryHandler - hMailServer has run out of memory, clearing caches."

That sound promising, btw, are you active on the hms forums as well? 
and if so, Is it fair to say, all topics below are somehow related?

https://www.hmailserver.com/forum/viewtopic.php?f=7&t=36754
https://www.hmailserver.com/forum/viewtopic.php?f=7&t=42197
https://www.hmailserver.com/forum/viewtopic.php?f=7&t=42249
https://www.hmailserver.com/forum/viewtopic.php?f=7&t=42584

**Comment by RvdHout** [comment]
@hunterius-prime Don't know if you finished testing and confirmed it is working properly, but once completed are you planning on making a pull request for above mentioned fix? 
If not, are you willing to share the 'fix' you created so it possibly can be committed/embedded into Experimental build(s)?

**Comment by hunterius-prime** [comment]
this fix is confirmed working and I have fixed issue with MIME part of rfc822 message, and now I am trying to make older 32bit build ( 5.6.9.) Some progress has been made and now we are deciding what would be the best scenario to share changes, 
so while in our production is running your -mod version, I applied and tested changes to 5.7. . I was considering to create pull request here ,and connect also your sources and create pull request there, when I will have all working under my account.
The fastest way(and now for me the easiest) is providing you with changes and you can include them into your build ( if so, then I would need built files asap :-) to be able deploy them) 
so you decide

**Comment by RvdHout** [comment]
Maybe it is easiest to simply fork 5.7 ([master branche](https://github.com/hmailserver/hmailserver/tree/master)) in your github repositories, and commit changes to that fork, you can then even create a pull request for the official build to be committed once and if martin returns (which seems rather unlikely atm)

I suspect changes are not that different between 5.6 (32-bit) and 5.7 (64-bit), so i should be able to filter out the additions/changes fairly easy

**Comment by hunterius-prime** [comment]
until then, here are changed files
[5.6.9-mod.zip](https://github.com/user-attachments/files/17413178/5.6.9-mod.zip)

changes are rather simple , checking requested octetStart against buffer size, 
and second checking if we are requesting MIME of message part

**Comment by RvdHout** [comment]
I'll check it out later today, thx

**Comment by RvdHout** [comment]
In IMAPFetch.cpp, line 416/417 you have defined 2 variables that is not used anywhere, leftovers from testing?
        
```
int start = iByteStart;
int buffSize = body.GetLength();
```

**Comment by hunterius-prime** [comment]
> In IMAPFetch.cpp, line 416/417 you have defined 2 variables that are not used anywhere as far as i can see, leftovers from testing?
> 
> ```
> int start = iByteStart;
> int buffSize = body.GetLength();
> ```

yes

**Comment by hunterius-prime** [comment]
I have reproduced another scenario that lead to same error, I will debug/fix/describe it later (I have fork of hmailserver already and working under branch 5.6.9)

**Comment by hunterius-prime** [comment]
Update:  BODY.PEEK[] ( without specifying body part, with "start" outside of message size boundary) will also trigger Error Code: HM5136 and  HM4208.
Reason: execution of this command has separate way of message read so previous fix does not apply.

Fix is in production testing , if no error will be found after Monday, than I will consider it for fixed
Changes are in my fork of 5.6.9 .

**Comment by RvdHout** [comment]
Will have a look, I am out the door right now
But i already see you changed the overall formatting of the file, which, unfortunately makes finding diffs is a whole lot harder🥲

**Comment by RvdHout** [comment]
I know It is a pain in the ass, have been struggling with the way the project is formatted in the past as well,

The author used these settings:
https://www.hmailserver.com/forum/viewtopic.php?p=214033#p214033
> In Visual Studio, the settings are under Tools ->Options->Text Editor->C/C++->Tabs where I have always used "Insert spaces" and tab size/indent size set to 3.

**Comment by ner00** [comment]
Is it possible that this fix, or some other commit since, caused the issue reported in https://github.com/roundcube/roundcubemail/issues/9011, where EML attachments cause the mail client to receive incorrect data from hMailServer?

I've been running hMailServer _5.6.9-B2641.64_ since October 2023 without the issue linked above, but later updates, namely _5.6.9-B2641.77-OpenSSL-3.0.15-Experimental_ and _5.6.9-B2641.81-OpenSSL-3.0.15-Experimental_ bring up the EML attachment problem.

**Comment by RvdHout** [comment]
That sound unlikely, as part of this fix was to fix the Roundcube issue you refer to, I haven't had a issue since, and now you claim it caused it.

Anyway without further info nothing we can do.
I tried various clients to send a mail with another eml as attachment, thunderbird, outlook, Roundcube itself through the markasspam plugin all seem to work

**Comment by ner00** [comment]
v5.6.9-B2641.64 works fine with the same environment, later revisions (.77 and .81) do not.

**Comment by ner00** [comment]
> part of this fix was to fix the Roundcube issue you refer to

I was under the impression that this fix was about fixing the memory allocation, not EML attachments breaking.

**Comment by RvdHout** [comment]
than obviously you did not read anything posted above

Anyway I don't get any of those errors...
What PHP version, what roundcube version? Could it be a xampp/apache thingy? all required php extensions loaded?

Did you try with a very simple email message?

**Comment by ner00** [comment]
You're right. I'll do more testing with multiple versions and bring the issue to roundcube team with debug logs so that we can firstly rule out that the issue is not my environment or their mail client itself. I'll follow up from there. Thanks for the suggestions, I'll come back to this when ready.

**Comment by ner00** [comment]
The issue I was experiencing previously between Roundcube and hMailServer, as described in https://github.com/roundcube/roundcubemail/issues/9011#issuecomment-2641264632, was a mixed bag. On the one hand hMailServer didn't seem to be replying the way Roundcube expected, on the other hand, Roundcube wasn't trying to mitigate the content of the (unusual?) FETCH reply by hMailServer.

I knew that not all clients (web, desktop, mobile) were having issue parsing the message from hMailServer, but usually the discussion piles on which project has the onus of implementing a fix, fortunately for me in this case one can already be found in https://github.com/roundcube/roundcubemail/commit/8f34bf9e6493b186c64e4ec5b2f97483c14c0e34

**Comment by martinknafve** [comment]
A bit hard to follow this issue, but it sounds like it's related to !459 which is being fixed. I'll close this as duplicate.
